src/time_bucket_ng.c

```diff
@@ -0,0 +1,154 @@
+		result = DatumGetDateADT(DirectFunctionCall3(ts_time_bucket_ng,
+													 interval,
+													 DateADTGetDatum(ts_date),
+													 DateADTGetDatum(origin)));
+	}
+	else
+	{
+		result = DatumGetDateADT(
+			DirectFunctionCall2(ts_time_bucket_ng, interval, DateADTGetDatum(ts_date)));
+	}
+
+	return DirectFunctionCall1(date_timestamp, DateADTGetDatum(result));
+}
+
+TSDLLEXPORT Datum
+ts_time_bucket_ng_timestamptz(PG_FUNCTION_ARGS)
+{
+	DateADT result;
+	Datum interval = PG_GETARG_DATUM(0);
+	DateADT ts_date = DatumGetDateADT(DirectFunctionCall1(timestamptz_date, PG_GETARG_DATUM(1)));
+
+	if (PG_NARGS() > 2)
+	{
+		DateADT origin = DatumGetDateADT(DirectFunctionCall1(timestamptz_date, PG_GETARG_DATUM(2)));
+		result = DatumGetDateADT(DirectFunctionCall3(ts_time_bucket_ng,
+													 interval,
+													 DateADTGetDatum(ts_date),
+													 DateADTGetDatum(origin)));
+	}
+	else
+	{
+		result = DatumGetDateADT(
+			DirectFunctionCall2(ts_time_bucket_ng, interval, DateADTGetDatum(ts_date)));
+	}
+
+	return DirectFunctionCall1(date_timestamptz, DateADTGetDatum(result));
+}
+
+TSDLLEXPORT Datum
+ts_time_bucket_ng(PG_FUNCTION_ARGS)
+{
+	Interval *interval = PG_GETARG_INTERVAL_P(0);
+	DateADT date = PG_GETARG_DATEADT(1);
+	DateADT origin_date = 0; // 2000-01-01
+	int origin_year = 2000, origin_month = 1, origin_day = 1;
+	int year, month, day;
+	int delta, bucket_number;
+
+	if ((interval->time != 0) || ((interval->month != 0) && (interval->day != 0)))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("unsupported interval: use either days and weeks, or months and years")));
+	}
+
+	if ((interval->month == 0) && (interval->day == 0))
+	{
+		/*
+		 * This will be fixed in furhter versions of ts_time_bucket_ng().
+		 * The reason why it's not yet implemented is that we want to start
+		 * experimenting with variable-sized buckets as soon as possible.
+		 * We know that fixed-sized buckets work OK and adding corresponding
+		 * logic will be trivial.
+		 */
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("unsupported interval: at least one day expected")));
+	}
+
+	if (PG_NARGS() > 2)
+	{
+		origin_date = PG_GETARG_DATUM(2);
+		j2date(origin_date + POSTGRES_EPOCH_JDATE, &origin_year, &origin_month, &origin_day);
+	}
+
+	if ((origin_day != 1) && (interval->month != 0))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("Invalid 'origin' day, 1st day of the month expected")));
+	}
+
+	if (DATE_NOT_FINITE(date))
+		PG_RETURN_DATEADT(date);
+
+	if (interval->month != 0)
+	{
+		/* Handle months and years */
+
+		j2date(date + POSTGRES_EPOCH_JDATE, &year, &month, &day);
+
+		if ((year < origin_year) || ((year == origin_year) && (month < origin_month)))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("date is before origin, please choose an origin that is before all "
+							"dates")));
```

Comment:
Suggesting different error code and message:

```suggestion
			ereport(ERROR,
					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
					 errmsg("origin must be before the given date")));
```